feat(root): v3 migration guidance (#1369)
mi6/ic-design-system: src/content/structured/get-started/migrating-to-v3.mdx

---
path: "/get-started/migrating-to-v3"

navPriority: 4

date: "2025-02-13"

title: "Migrating to V3"

subTitle: "This guide is intended to help you migrate from ICDS 2.x components to ICDS 3.x components."

contribute: "https://github.com/mi6/ic-design-system/tree/main/src/content/structured/get-started/migrating-to-v3.mdx"
---

## Introduction

> ICDS V3 introduces one of our mostly highly anticipated features, dark mode as well as general improvements. This is a comprehensive guide on how to upgrade your V2 app to V3. For a full list of the changes made in V3, please read the [changelog](https://github.com/mi6/ic-ui-kit/blob/develop/packages/web-components/CHANGELOG.md).

Reviewer: Should this link have the open in new window icon?

> ICDS V3 introduces one of our mostly highly anticipated features, dark mode as well as general improvements. This is a comprehensive guide on how to upgrade your V2 app to V3. For a full list of the changes made in V3, please read the [changelog](https://github.com/mi6/ic-ui-kit/blob/develop/packages/web-components/CHANGELOG.md).

Reviewer 2: Add comma after dark mode: ICDS V3 introduces one of our most highly anticipated features, dark mode, as well as general improvements.

### 2. Component renaming

- `ic-card` (`IcCard`) has been renamed to `ic-card-vertical` (`IcCardVertical`) .
- `ic-data-entity` (`IcDataEntity`) has been renamed to `ic-data-list` (`IcDataList`).

### 3. Removed props/slots/events/methods

- Deprecated props, slots, events, and methods have been removed entirely.
- The `maxLength` prop has been removed. Instead, use the `maxCharacters` prop and the `hideCharCount` prop to hide the character count visually.
- The `small` prop has been removed. Please use `size="small"` instead.
- The `size=”default”` prop has been removed. Please use `size="medium"` instead.

### 4. Class name changes

- Class names on the host for most components have been updated to include the component name in the class name. For example, `disabled` has been updated to `ic-select-disabled`.

### 5. Event name changes

- The `icThemeChange` event has been renamed to `icBrandChange`.
- The events `sideNavExpanded` and `topNavResized` are now called `icSideNavExpanded` and `icTopNavResized`.

### 6. CSS token changes

- Some existing CSS tokens have been renamed following the move towards primitive, semantic, and component tokens.

> ### 7. Component behavior changes

Reviewer 2: Change behavior to behaviour.

- The `ic-toast` (`IcToast`) component now shows a dismiss button when the toast is hovered over or focused.

> - The shadow DOM has been removed from `ic-radio` (`IcRadio`) and may affect developers current tests.

Reviewer 2: Either change to 'may affect developer's current tests' or just remove developers so it's 'may affect current tests'.

### How it works

Given a directory, it will scan over files and find any relevant ICDS components that require a change either by component name or props used.

### How to run

<p>
  This package works as an executable requiring a directory and optional test
  boolean argument to help migrate your tests from ICDS V2 to ICDS V3. Further
  information and examples can be found in the{" "}
  <ic-link
    target="_blank"
    href="https://github.com/mi6/ic-ui-kit/blob/v3.0.0/main/packages/codemod/README.md"
    rel="noreferrer noopener"
  >
    codemod package README
  </ic-link>
  .
</p>

```console
- npm @ukic/react@latest
- npm @ukic/web-components@latest
- npx @ukic/codemod <dir> <test>
```

## Breaking changes

### 1. Prop renaming

- The `appearance` prop has been replaced with the `theme` and `monochrome` [props](/get-started/development/custom-theme/#dark-mode-as-a-theme).
- The `auto-focus` prop has been renamed to `autofocus`.

> - The `color` prop has been renamed to `brandColor`.

Reviewer 3: Please add what components this affects

- The `group-title` prop in `ic-accordion-group` (`IcAccordionGroup`) and `ic-footer-link-group` (`IcFooterLinkGroup`) has been renamed to `label`.
- The `text-label` prop in `ic-badge` (`IcBadge`) has been renamed to `label`.
- The `body-max-lines` prop in `ic-empty-state` (`IcEmptyState`) has been renamed to `max-lines`.
- The `error` prop in `ic-input-label` (`IcInputLabel`) has been renamed to `status` with `error` as an option.
- The `boundary-count` and `adjacent-count` props in `ic-pagination` (`IcPagination`) have been renamed to `boundary-page-count` and `adjacent-page-count`.
- The `disable-filter` prop in `ic-search-bar` (`IcSearchBar`) and `ic-select` (`IcSelect`) has been renamed to `disable-auto-filtering`.
- The `hint-text` prop in `ic-search-bar` (`IcSearchBar`) has been renamed to `assistive-hint-text`.

> - Some `ic-step` (`IcStep`) props have had the prefix "step" removed, e.g., `step-status` has been renamed to `status`.

Reviewer 3: Please add all the prop changes in IcStep as well as any missing props mentioned in the [other migration PR](https://github.com/mi6/ic-design-system/pull/1264/files#diff-064a856c0f72ab51a0d8801a34b9bfdfd414915d9c81976ff2cc1beede7f4aba). I've seen that the `keyboardShortcut` prop rename in `ic-menu-item` is missing but haven't checked if any others are missing

## Codemod

To aid your migration from ICDS V2 to V3, we have created a codemod to automate the process of migrating your code. A codemod is a tool that can be used to make large-scale changes to codebases quickly and efficiently. Our ICDS V2 to V3 codemod is the simplest and easiest way to migrate your codebase to the latest version of the UI Kit.

### How it works

Given a directory, it will scan over files and find any relevant ICDS components that require a change either by component name or props used.

### How to run

<p>
  This package works as an executable requiring a directory and optional test
  boolean argument to help migrate your tests from ICDS V2 to ICDS V3. Further
  information and examples can be found in the{" "}
  <ic-link
    target="_blank"
    href="https://github.com/mi6/ic-ui-kit/blob/v3.0.0/main/packages/codemod/README.md"
    rel="noreferrer noopener"
  >
    codemod package README
  </ic-link>
  .
</p>

```console

> - npm @ukic/react@latest

Reviewer: Should this be `npm i ...` or have I misunderstood and its doing something different?

  label="What are your favourite types of coffee?"
  multiple="true"
></ic-select-with-multi>
// v3
<ic-select
  id="select-default"
  label="What are your favourite types of coffee?"
  multiple="true"
></ic-select>
```

## New features

### 1. Theme prop

- A new `theme` prop has been added to many components to enable light and dark mode.

### 2. Monochrome prop

- A new `monochrome` prop has been added to components like `ic-toggle-button` (`IcToggleButton`) and `ic-toggle-button-group` (`IcToggleButtonGroup`).

## Manual migration steps

1. **Update Prop Names**: Replace all instances of the old prop names with the new ones.
2. **Update Component Names**: Rename components according to the new names.
3. **Remove Deprecated Props/Slots/Events/Methods**: Remove any usage of deprecated props, slots, events, and methods.
4. **Update Class Names**: Update class names to the new format.
5. **Update Event Names**: Replace old event names with the new ones.
6. **Update CSS Tokens**: Update any CSS tokens that have been renamed.
7. **Test Your Application**: Thoroughly test your application to ensure that all changes have been applied correctly and that there are no regressions.

> By following these steps, you should be able to successfully migrate your project from version 2 to version 3 of the `@ukic/web-components` package.

Reviewer 4: Should this be either package?
```suggestion
By following these steps, you should be able to successfully migrate your project from version 2 to version 3 of the `@ukic/web-components` or `@ukic/react` package.
```

---
path: "/get-started/migrating-to-v3"

navPriority: 4

date: "2025-02-13"

title: "Migrating to v3"

subTitle: "Developer guidance on migration from v2 to v3 of the UI-Kit"

contribute: "https://github.com/mi6/ic-design-system/tree/main/src/content/structured/get-started/development/releases-versioning.mdx"
---

> 

Reviewer 3: Please add this paragraph (taken from https://github.com/mi6/ic-design-system/pull/1264/files#diff-064a856c0f72ab51a0d8801a34b9bfdfd414915d9c81976ff2cc1beede7f4abaR10-R13)
## Introduction

ICDS V3 introduces one of our mostly highly anticipated features, dark mode as well as general improvements. This is a comprehensive guide on how to upgrade your V2 app to V3. For a full list of the changes made in V3, please read the <ic-link href="https://github.com/mi6/ic-ui-kit/tree/v3.0.0/main/packages/web-components">3.x changelog</ic-link>.

> ## Ic-theme and dark mode

Reviewer: ```suggestion
## Dark mode and ic-theme
```

Don't like the capitalisation of the component name there, so could maybe swap them around for this heading?

> With v3's arrival comes the introduction of the `ic-theme` component. This is responsible for managing the theme of the UI Kit and providing a dark mode. Guidance on using the `ic-theme` package can be found in the [INSERT IC-THEME DOCUMENTATION](/theme).

Reviewer: ```suggestion
With v3's arrival comes the introduction of the `ic-theme` component. This is responsible for managing the theme of the UI Kit and providing a dark mode. Guidance on implementing `ic-theme` can be found in the [custom theme guidance](/get-started/development/custom-theme).
```

> ## Ic-theme and dark mode
> 
> With v3's arrival comes the introduction of the `ic-theme` component. This is responsible for managing the theme of the UI Kit and providing a dark mode. Guidance on using the `ic-theme` package can be found in the [INSERT IC-THEME DOCUMENTATION](/theme).

Reviewer 3: ```suggestion
## Theming and dark mode

With V3's arrival comes an improved `ic-theme` component. This component was initially responsible for managing the theme of the UI Kit. In addition to the existing functionality, you can now use the `ic-theme` component to set whether a component responds to changes in the system's light and dark theme or whether it always appears in a particular mode. Guidance on how to use the `ic-theme` package can be found in the [INSERT IC-THEME DOCUMENTATION](/theme)
```

## Codemod

> Codemod is a script used to automate the process of migrating code from one version to another. It is a tool that can be used to make large-scale changes to codebases quickly and efficiently.It is the simplest and easiest way to migrate your codebase to the latest version of the UI Kit.

Reviewer: ```suggestion
Codemod is a script used to automate the process of migrating code from one version to another. It is a tool that can be used to make large-scale changes to codebases quickly and efficiently. It is the simplest and easiest way to migrate your codebase to the latest version of the UI Kit.
```

> Codemod is a script used to automate the process of migrating code from one version to another. It is a tool that can be used to make large-scale changes to codebases quickly and efficiently.It is the simplest and easiest way to migrate your codebase to the latest version of the UI Kit.

Reviewer 3: ```suggestion
To aid your migration from ICDS V2 to V3, we have created a codemod to automate the process of migrating your code. A codemod is a tool that can be used to make large-scale changes to codebases quickly and efficiently. Our ICDS V2 to V3 codemod is the simplest and easiest way to migrate your codebase to the latest version of the UI Kit.
```

> ### What is this used for

Reviewer 3: ```suggestion
### How it works
```

Given a directory, it will scan over files and find any relevant ICDS components that require a change either by component name or props used.

### How to run

> This package will be usable as an executable requiring a directory and optional test boolean argument to cover tests. Further information can be found in the codemod package [README](https://github.com/mi6/ic-ui-kit/blob/v3.0.0/main/packages/codemod/README.md).

Reviewer 3: ```suggestion
This package works as an executable requiring a directory and optional test boolean argument to help migrate your tests from ICDS V2 to ICDS V3. Further information can be found in the [codemod package README](https://github.com/mi6/ic-ui-kit/blob/v3.0.0/main/packages/codemod/README.md).
```

```console
- npx @ukic/codemod <dir> <test>
```

> ## Breaking Changes - web-components

Reviewer: Is there going to be a heading for React? If not could the web components bit be removed? Also:
```suggestion
## Breaking changes - web-components
```

### 1. Prop Renaming

> - The `appearance` prop has been replaced with the `theme` and `monochrome` props.

Reviewer: ```suggestion
- The `appearance` prop has been replaced with the `theme` and `monochrome` [props](/get-started/development/custom-theme/#dark-mode-as-a-theme).
```

- The `auto-focus` prop has been renamed to `autofocus`.
- The `color` prop has been renamed to `brandColor`.
- The `icThemeChange` event has been renamed to `icBrandChange`.

### 2. Component Renaming

> - `ic-card` has been renamed to `ic-card-vertical`.

Reviewer 3: ```suggestion
- `ic-card` (`IcCard`) has been renamed to `ic-card-vertical` (`IcCardVertical`) .
```

> - `ic-data-entity` has been renamed to `ic-data-list`.

Reviewer 3: ```suggestion
- `ic-data-entity` (`IcDataEntity`)  has been renamed to `ic-data-list` (`IcDataList`).
```

### 3. Removed Props/Slots/Events/Methods

> - Deprecated props, slots, events, and methods have been removed entirely.
> - The `maxLength` prop has been removed. Instead, use the `maxCharacters` prop and the `hideCharCount` prop to hide the character count visually.

Reviewer 3: ```suggestion
- Deprecated props, slots, events, and methods have been removed entirely.
- The `maxLength` prop has been removed. Instead, use the `maxCharacters` prop and the `hideCharCount` prop to hide the character count visually.
- The `small` prop has been removed. Please use `size="small"` instead.
- The `size=”default”` prop has been removed. Please use `size="medium"` instead.
```

### 4. Class Name Changes

- Class names on the host for most components have been updated to include the component name in the class name. For example, `disabled` has been updated to `ic-select-disabled`.

### 5. Event Name Changes

- The events `sideNavExpanded` and `topNavResized` are now called `icSideNavExpanded` and `icTopNavResized`.

### 6. CSS Token Changes

- Some existing CSS tokens have been renamed following the move towards primitive, semantic, and component tokens.

### 7. Component Behavior Changes

> - The `ic-select` component now needs to be used to get the multi-select variant (use the `multiple` prop), instead of `ic-select-with-multi` from `canary-web-components`.
> - The `ic-toast` component now shows a dismiss button when the toast is hovered over or focused.

Reviewer 3: ```suggestion
- The `ic-toast` component now shows a dismiss button when the toast is hovered over or focused.
-`ic-multi-select` (`IcMultiSelect`) is no longer a standalone component. The multi-select functionality can now be achieved using the `ic-select` (`IcSelect`) component and its `multiple` prop.

```tsx
// v2
<ic-select-with-multi
  id="select-default"
  label="What are your favourite types of coffee?"
  multiple="true"
></ic-select-with-multi>

// v3
<ic-select
  id="select-default"
  label="What are your favourite types of coffee?"
  multiple="true"
></ic-select>
```
```
